File: python_src/hgtm/task_migration_by_groups.py
"""TaskMigrationByGroups class for group-based task migration"""
import sys
sys.path.append('..')
from input.migration_record import MigrationRecord
from mpftm.task_migration_based_pon import TaskMigrationBasedPon


class TaskMigrationByGroups:
    def __init__(self, arc_graph, id_to_groups, id_to_agents,
                 group_id_to_pfield, robot_id_to_pfield,
                 records, a, b, id_to_i):
        self.arc_graph = arc_graph
        self.id_to_groups = id_to_groups
        self.id_to_agents = id_to_agents
        self.group_id_to_pfield = group_id_to_pfield
        self.robot_id_to_pfield = robot_id_to_pfield
        self.records = records
        self.a = a
        self.b = b
        self.id_to_i = id_to_i

    def run(self, bags_to_agent):
        """Run task migration by groups"""
        rece_agents = []

        for bag, agent in bags_to_agent.items():
            if agent is None:
                continue

            q_size = len(agent.get_tasks_list()) if agent.get_tasks_list() else 0

            g_size = sum(len(e.get_tasks_list()) for e in bag)

            group_id = agent.get_group_id()
            rl = self.id_to_groups[group_id].get_interaction_level()

            if g_size * (1 - rl) * 2 > q_size:
                rece_agents.append(agent)

        migration_agents = []

        for agent_id in self.id_to_agents.keys():
            agent = self.id_to_agents[agent_id]
            if agent.get_fault_a() == 1:
                migration_agents.append(agent)
                agent.set_fault_a(0)
                agent.set_fault_o(1)

        for rece_agent in rece_agents:
            rece_agent.set_fault_a(1)

        # Pre-execute MPFTM algorithm
        migration_records = TaskMigrationBasedPon(
            self.id_to_groups, self.id_to_agents, self.arc_graph,
            self.group_id_to_pfield, self.robot_id_to_pfield,
            self.id_to_i, self.a, self.b
        ).run()

        self.records.extend(migration_records)

        for rece_agent in rece_agents:
            rece_agent.set_fault_a(0)

        for bag, agent_migrated in bags_to_agent.items():
            for agent in bag:
                for task in list(agent.get_tasks_list()):
                    self.execute_migration(agent, agent_migrated, task)

        return self.records

    def execute_migration(self, agent, agent_migrated, migration_task):
        """Execute migration"""
        if agent is None or agent_migrated is None or migration_task is None:
            return

        # Update inter-layer load and task list
        if agent.get_group_id() != agent_migrated.get_group_id():
            self.update_inter(agent, agent_migrated, migration_task)

        # Update intra-layer load and task list
        self.update_intra(agent, agent_migrated, migration_task)

        record = MigrationRecord()
        record.set_from(agent.get_robot_id())
        record.set_to(agent_migrated.get_robot_id())
        self.records.append(record)

    def update_inter(self, robot, robot_migrated, migration_task):
        """Update inter-layer migration"""
        if robot is None or robot_migrated is None or migration_task is None:
            return

        group_id = robot.get_group_id()
        group = self.id_to_groups[group_id]

        robot_migrated_group_id = robot_migrated.get_group_id()
        migrated_group = self.id_to_groups[robot_migrated_group_id]

        group.set_group_load(group.get_group_load() - migration_task.get_size())
        migrated_group.set_group_load(migrated_group.get_group_load() + migration_task.get_size())

    def update_intra(self, robot, robot_migrated, migration_task):
        """Update intra-layer migration"""
        tasks_list = robot.get_tasks_list()
        tasks_list.remove(migration_task)
        robot.set_load(robot.get_load() - migration_task.get_size())
        robot.set_tasks_list(tasks_list)

        robot_migrated_task_list = robot_migrated.get_tasks_list()
        if robot_migrated_task_list is None:
            robot_migrated_task_list = []

        robot_migrated_task_list.append(migration_task)
        robot_migrated.set_load(robot_migrated.get_load() + migration_task.get_size())
        robot_migrated.set_tasks_list(robot_migrated_task_list)

File: src/main/java/HGTM/TaskMigrationByGroups.java
package HGTM;

import MPFTM.TaskMigrationBasedPon;
import input.*;
import org.jgrapht.alg.interfaces.ShortestPathAlgorithm;
import org.jgrapht.graph.DefaultUndirectedWeightedGraph;
import org.jgrapht.graph.DefaultWeightedEdge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class TaskMigrationByGroups {
    private DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph;
    private HashMap<Integer, Group> idToGroups;
    private HashMap<Integer, Agent> idToAgents;
    private ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath ;
    private HashMap<Integer, PotentialField> groupIdToPfield;
    private HashMap<Integer,PotentialField> robotIdToPfield;
    private List<MigrationRecord> records;
    Double a;
    Double b;
    private HashMap<Integer,Double> idToI;

    public TaskMigrationByGroups(DefaultUndirectedWeightedGraph<Integer, DefaultWeightedEdge> arcGraph, HashMap<Integer, Group> idToGroups, HashMap<Integer, Agent> idToAgents,
                                 ShortestPathAlgorithm<Integer, DefaultWeightedEdge> shortestPath, HashMap<Integer, PotentialField> groupIdToPfield, HashMap<Integer, PotentialField> robotIdToPfield,
                                 List<MigrationRecord> records, Double a, Double b, HashMap<Integer, Double> idToI) {
        this.arcGraph = arcGraph;
        this.idToGroups = idToGroups;
        this.idToAgents = idToAgents;
        this.shortestPath = shortestPath;
        this.groupIdToPfield = groupIdToPfield;
        this.robotIdToPfield = robotIdToPfield;
        this.records = records;
        this.a = a;
        this.b = b;
        this.idToI = idToI;
    }
    public List<MigrationRecord> run(Map<List<Agent>,Agent> bagsToAgent) {
        List<Agent> receAgents = new ArrayList<>();
        for (List<Agent> bag : bagsToAgent.keySet()) {
            Agent agent = bagsToAgent.get(bag);
            if (agent==null) {
                continue;
            }
            int Qsize = 0;
            if (agent.getTasksList() != null) {
                Qsize = agent.getTasksList().size();
            }
            int Gsize = 0;
            for (Agent e : bag) {
                Gsize+=e.getTasksList().size();
            }
            int groupId = agent.getGroupId();
            double RL = idToGroups.get(groupId).getInteractionLevel();
            if (Gsize*(1-RL)*2>Qsize) {
                receAgents.add(agent);
            } else {
                //bagsToAgent.remove(bag);
            }
        }
        List<Agent> MigrationAgents = new ArrayList<>();
        for (Integer id : idToAgents.keySet()) {
            Agent agent = idToAgents.get(id);
            if (agent.getFaultA()==1) {
                MigrationAgents.add(agent);
                agent.setFaultA(0);
                agent.setFaultO(1);
                //为了方便将接收任务的组件上原有的任务扩散出去，也为了能够更好的利用原有的MPTFM算法
            }
        }
        for (Agent receAgent : receAgents) {
            receAgent.setFaultA(1);
            //为了方便将接收任务的组件上原有的任务扩散出去，只有FaultA=1的组件上的任务才会被扩散出去
        }
        //对接收任务的组件提前执行MPFTM算法--腾出负载空间
        List<MigrationRecord> migrationRecords = new TaskMigrationBasedPon(idToGroups, idToAgents,
                arcGraph, groupIdToPfield, robotIdToPfield, shortestPath, idToI, a, b).run();
        records.addAll(migrationRecords);
        for (Agent receAgent : receAgents) {
            //还原接收任务的组件状态
            receAgent.setFaultA(0);
        }
        for (List<Agent> bag : bagsToAgent.keySet()) {
            Agent agentMigrated = bagsToAgent.get(bag);
            for (Agent agent : bag) {
                for (Task task : agent.getTasksList()) {
                    excuteMigration(agent,agentMigrated,task);
                }
            }
        }

        return records;
    }

    private void excuteMigration(Agent agent, Agent agentMigrated, Task migrationTask) {
        if (agent == null) {
            return;
        }

        if (agentMigrated == null) {
            return;
        }
        if (migrationTask == null) {
            return;
        }

        //改变网络层间负载和任务列表
        if (agent.getGroupId()!=agentMigrated.getGroupId()) {
            //不在同一组才会更新网络层间的势场情况
            updateInter(agent, agentMigrated, migrationTask);
        }
        //改变网络层内负载和任务列表
        updateIntra(agent, agentMigrated, migrationTask);
        MigrationRecord record = new MigrationRecord();
        record.setFrom(agent.getRobotId());
        record.setTo(agentMigrated.getRobotId());
        records.add(record);
        //初始化（计算）上下文负载

    }
    private void updateInter(Agent robot, Agent robotMigrated, Task migrationTask) {
        if (robot == null) {
            return;
        }

        if (robotMigrated == null) {
            return;
        }
        if (migrationTask == null) {
            return;
        }

        int groupId = robot.getGroupId();
        Group group = idToGroups.get(groupId);
        int robotMigratedGroupId = robotMigrated.getGroupId();
        Group migratedGroup = idToGroups.get(robotMigratedGroupId);

        group.setGroupLoad(group.getGroupLoad()-migrationTask.getSize());
        migratedGroup.setGroupLoad(migratedGroup.getGroupLoad()+migrationTask.getSize());


    }

    private void updateIntra(Agent robot, Agent robotMigrated, Task migrationTask) {
        //更新迁移后的任务负载
        List<Task> tasksList = robot.getTasksList();
        tasksList.remove(migrationTask);
        robot.setLoad(robot.getLoad()-migrationTask.getSize());
        robot.setTasksList(tasksList);


        List<Task> robotMigratedTaskList = robotMigrated.getTasksList();
        if (robotMigratedTaskList==null) {
            robotMigratedTaskList = new ArrayList<>();
        }

        robotMigratedTaskList.add(migrationTask);
        robotMigrated.setLoad(robotMigrated.getLoad()+migrationTask.getSize());
        robotMigrated.setTasksList(robotMigratedTaskList);

    }
}
